### PR TITLE
ENH Various changes to support SiteTree form field scaffolding

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- https://github.com/silverstripe/silverstripe-cms/issues/2767

### CODE REVIEW

File: src/Forms/FormScaffolder.php [hunk 1]
```diff
@@ -35,32 +35,26 @@ class FormScaffolder
...
      */
-    public $fieldClasses;
+    public array $fieldClasses = [];
 
     /**
-     * @var boolean $includeRelations Include has_many and many_many relations
+     * Include has_many and many_many relations
      */
-    public $includeRelations = false;
+    public bool|array $includeRelations = false;
```
Comment (GuySartorelli) [inline src/Forms/FormScaffolder.php:57]: Looking at the code that uses this, you could use it to explicitly only allow `has_many` relations but not `many_many` relations (or vice versa) - in which case it needs to be an associative array. Hence `bool|array`.

File: src/Forms/FormScaffolder.php
Comment (GuySartorelli) [inline src/Forms/FormScaffolder.php:1]: Added strict typing for the properties as a matter of course - it's unrelated refactoring.

File: src/Forms/FormScaffolder.php [hunk 2]
```diff
@@ -106,15 +100,15 @@ public function getFieldList()
         // Add logical fields directly specified in db config
         foreach ($this->obj->config()->get('db') as $fieldName => $fieldType) {
             // Skip fields that aren't in the allow list
-            if ($this->restrictFields && !in_array($fieldName, $this->restrictFields ?? [])) {
+            if (!empty($this->restrictFields) && !in_array($fieldName, $this->restrictFields)) {
```
Comment (GuySartorelli) [inline src/Forms/FormScaffolder.php:103]: Must be an array, so we can tidy this a little now. Applies to similar changes below as well.

File: src/Forms/FormScaffolder.php [hunk 3]
```diff
@@ -106,15 +100,15 @@ public function getFieldList()
...
                 continue;
             }
             // Skip ignored fields
             if (in_array($fieldName, $this->ignoreFields)) {
                 continue;
             }
 
-            if ($this->fieldClasses && isset($this->fieldClasses[$fieldName])) {
+            if (isset($this->fieldClasses[$fieldName])) {
```
Comment (GuySartorelli) [inline src/Forms/FormScaffolder.php:111]: Must be an array, so we can tidy this a little now.

File: src/Forms/Tab.php [hunk 4]
```diff
@@ -99,6 +98,12 @@ public function setID($id)
         return $this;
     }
 
+    public function setName($name)
+    {
+        $this->setID(Convert::raw2htmlid($name));
+        return parent::setName($name);
+    }
```
Comment (GuySartorelli) [inline src/Forms/Tab.php:105]: This ensures the `id` attribute in the HTML markup correctly reflects the name, even if the name is set after the tab was created. This is needed by https://github.com/silverstripe/silverstripe-userforms/pull/1315/files#diff-d659fc65ecb263c36d5428815131e80ef3c1750bf18e18860f1fd9d79d4df3e0R195 because the relation name doesn't match the name we want for the tab.

Comment (GuySartorelli) [reply src/Forms/Tab.php:105]: It's also just a sensible change to make.

File: src/Forms/Tab.php [hunk 5]
```diff
@@ -99,6 +98,14 @@ public function setID($id)
         return $this;
     }
 
+    public function setName($name)
+    {
+        if ($this->ID() === Convert::raw2htmlid($this->getName())) {
+            $this->setID(Convert::raw2htmlid($name));
+        }
+        return parent::setName($name);
+    }
```
Comment (GuySartorelli) [inline src/Forms/Tab.php]: This ensures the `id` attribute in the HTML markup correctly reflects the name, even if the name is set after the tab was created. This is needed by https://github.com/silverstripe/silverstripe-userforms/pull/1315/files#diff-d659fc65ecb263c36d5428815131e80ef3c1750bf18e18860f1fd9d79d4df3e0R195 because the relation name doesn't match the name we want for the tab.

It's also just a sensible change to make.

Note that we don't set the ID if the ID was manually set to something other than the name.

File: src/ORM/DataObject.php
Comment (GuySartorelli) [inline src/ORM/DataObject.php:1]: Changes here are to support the new strict types, and removing the deprecated `ajaxSafe`

File: tests/php/Forms/TabTest.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        $tab->setName('NewName');
+        $this->assertSame('NewName', $tab->getName());
+        $this->assertSame('NewName', $tab->ID());
+
+        // If ID is explicitly set, changing the name doesn't override it
+        $tab->setID('Custom-ID');
+        $tab->setName('AnotherName');
+        $this->assertSame('AnotherName', $tab->getName());
+        $this->assertSame('Custom-ID', $tab->ID());
+    }
```
Comment (GuySartorelli) [inline tests/php/Forms/TabTest.php:29]: Can't use a data provider for this - the scenarios need to be run sequentially.

**Review by emteknetnz — APPROVED** [review]
